[NFC] Refactor ConstantArrayType size storage

llvm-beanz opened the pull request:
In PR #79382, I need to add a new type that derives from ConstantArrayType. This means that ConstantArrayType can no longer use `llvm::TrailingObjects` to store the trailing optional Expr*.

This change refactors ConstantArrayType to store a 60-bit integer and 4-bits for the integer size in bytes. This replaces the APInt field previously in the type but preserves enough information to recreate it where needed.

To reduce the number of places where the APInt is re-constructed I've also added some helper methods to the ConstantArrayType to allow some common use cases that operate on either the stored small integer or the APInt as appropriate.

Resolves #85124.

llvm-beanz committed:
[NFC] Refactor ConstantArrayType size storage

In PR #79382, I need to add a new type that derives from
ConstantArrayType. This means that ConstantArrayType can no longer use
`llvm::TrailingObjects` to store the trailing optional Expr*.

This change refactors ConstantArrayType to store a 60-bit integer and
4-bits for the integer size in bytes. This replaces the APInt field
previously in the type but preserves enough information to recreate it
where needed.

To reduce the number of places where the APInt is re-constructed I've
also added some helper methods to the ConstantArrayType to allow some
common use cases that operate on either the stored small integer or the
APInt as appropriate.

Resolves #85124.

tbaederr commented on clang/include/clang/AST/Type.h:
```diff
@@ -3182,34 +3182,98 @@ class ArrayType : public Type, public llvm::FoldingSetNode {
+      : ArrayType(ConstantArray, Et, Can, SM, TQ, nullptr), I(Width / 8, Sz) {
+    ConstantArrayTypeBits.HasExternalSize = false;
+    assert(Sz < 0x0FFFFFFFFFFFFFFF && "Size must fit in 60 bits");
+    assert(Width < 0xFF && "Type width must fit in 8 bits");
```
8 or 4?

llvm-beanz replied:
I guess that's poorly worded. I verify that Width is within 8, but I only store 4 since the least significant bits are always zero.

efriedma-quic replied:
You divide by 8, then truncate the bits above the lowest 4... so the max value you can actually store is 15*8==120.

llvm-beanz committed:
Update comment to clarify the assert.

efriedma-quic commented on clang/include/clang/AST/Type.h:
```diff
@@ -3182,34 +3182,100 @@ class ArrayType : public Type, public llvm::FoldingSetNode {
+               : I.Size;
+  }
+
+  /// Return the size zero-extended as a uint64_t.
```
Sign-extended

efriedma-quic commented on clang/include/clang/AST/Type.h:
```diff
@@ -3182,34 +3182,100 @@ class ArrayType : public Type, public llvm::FoldingSetNode {
+  };
+  struct InlineSize {
+    InlineSize(uint64_t TSz, uint64_t Sz) : ByteWidth(TSz), Size(Sz) {}
+    uint64_t ByteWidth : 4;
```
Do we actually need to store the ByteWidth here?  It's always just Target->getMaxPointerWidth().  Granted, it might be inconvenient to change the API... in which case, I guess we can store it here.

Maybe consider sticking the ByteWidth in ConstantArrayTypeBitfields; there should be a few spare bits available there.

efriedma-quic commented on clang/include/clang/AST/Type.h:
```diff
@@ -3182,34 +3182,100 @@ class ArrayType : public Type, public llvm::FoldingSetNode {
+  union {
+    struct InlineSize I;
+    ExternalSize *SizePtr;
+  };
```
If I'm reading this correctly, this actually saves 8 bytes over the current representation?  That's nice.

llvm-beanz committed:
Updates based on feedback from @efriedma-quic, thanks!

efriedma-quic reviewed (approved):
LGTM

llvm-beanz committed:
Update static_assert, remove FIXME

This just updates a static assert and removes an old FIXME since this
patch fixes it.

llvm-beanz committed:
clang-format

coopp reviewed (approved):
Looks good to me for what I can see in the code.

I took special care to make sure places where noticed the old size accessor (CAT->getSize().getZExtValue() ) being called, was now using (CAT->getZExtSize()).
